src/nccl_ofi_rdma.c

```diff
@@ -306,65 +263,77 @@ static void unlink_topo_file()
 {
 	int ret = 0;
-	int rc = 0;
-	FILE *file;
-	char *filename;
-	int fd;
+	int topo_fd = -1;
+	int write_fd = -1;
+	FILE *file = NULL;
 
 	/* This function is a no-op in case writing topology file is not enabled explicitly */
 	if (!ofi_nccl_topo_file_write_enable()) {
+		NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET,
+			       "Topology write not enabled; skipping");
 		goto exit;
 	}
 
-	nccl_net_ofi_mutex_lock(&topo_file_lock);
-
-	if (topo_file_unlink) {
-		/* A topology file has already been written and stored
-		 * such that it can be unlinked later. Do not write
-		 * another topology file since it would end up
-		 * overriding the stored filename. */
-		goto unlock;
+	/* If `NCCL_TOPO_FILE` is already set, don't set it again. */
+	if (getenv("NCCL_TOPO_FILE")) {
+		NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET,
+			       "NCCL_TOPO_FILE environment variable is already set to %s",
+			       getenv("NCCL_TOPO_FILE"));
+		goto exit;
 	}
 
-	if (ofi_nccl_topo_file_template()) {
-		filename = strdup(ofi_nccl_topo_file_template());
-	} else {
-		filename = strdup(topo_file_template);
-		/* Store filename to be unlinked later */
-		topo_file_unlink = filename;
+	/* Create file descriptor */
+	topo_fd = memfd_create("ofi_nccl_topo", 0);
+	if (topo_fd == -1) {
+		NCCL_OFI_WARN("Failed to create anonymous topology file. ERROR: %s",
+			      strerror(errno));
+		ret = -errno;
+		goto error;
 	}
 
-	/* Create file descriptor */
-	fd = mkstemp(filename);
-	if (fd == -1) {
-		NCCL_OFI_WARN("Failed to create NCCL topology file from template %s. ERROR: %s",
-			      filename, strerror(errno));
+	/* Duplicate the fd to open the file for writing */
+	write_fd = dup(topo_fd);
+	if (write_fd == -1) {
+		NCCL_OFI_WARN("Failed to create topo write fd. ERROR: %s",
+			      strerror(errno));
 		ret = -errno;
-		goto unlock;
+		goto error;
 	}
 
 	/* Open file from file descriptor */
-	file = fdopen(fd, "w");
+	file = fdopen(write_fd, "w");
 	if (file == NULL) {
-		NCCL_OFI_WARN("Failed to open NCCL topology file using file descriptor. File name: %s. ERROR %s",
-			      filename, strerror(errno));
+		NCCL_OFI_WARN("Failed to open NCCL topology file using file descriptor. ERROR %s",
+			      strerror(errno));
 		ret = -errno;
-		goto unlock;
+		goto error;
 	}
+	write_fd = -1;
 
 	ret = nccl_ofi_topo_write(topo, file);
 	if (ret) {
-		NCCL_OFI_WARN("Failed to write NCCL topology using file descriptor. File name: %s",
-			      filename);
-		goto unlock;
+		NCCL_OFI_WARN("Failed to write NCCL topology using file descriptor. RC: %d", ret);
+		goto error;
 	}
 
-	/* Close file. The file remains accessible as long as file is not unlinked. */
+	/* Close file. The file at /dev/fd/{topo_fd} remains accessible as long as the process lives. */
 	if (fclose(file) == EOF) {
-		NCCL_OFI_WARN("Unable to close NCCL topology file. File name: %s. ERROR: %s",
-			      filename, strerror(errno));
+		NCCL_OFI_WARN("Unable to close NCCL topology file. ERROR: %s",
```

Comment:
there's no point in closing the memfd file unless there was an error.

Reply:
writes should be flushed, though

Reply:
Yes, agreed.